src/lib.rs

```diff
@@ -29,38 +28,101 @@ mod dlmalloc;
 #[cfg(all(feature = "global", not(test)))]
 mod global;
 
+/// A platform interface for platforms that support the global lock
+#[cfg(feature = "global")]
+pub trait GlobalSystem: System {
+    /// Acquires the global lock
+    fn acquire_global_lock();
+
+    /// Releases the global lock
+    fn release_global_lock();
+}
+
+/// Type alias for the current platform
+#[cfg(target_os = "linux")]
+pub type Platform = Linux;
+
+/// Type alias for the current platform
+#[cfg(target_os = "macos")]
+pub type Platform = MacOS;
+
+/// Type alias for the current platform
+#[cfg(target_arch = "wasm32")]
+pub type Platform = Wasm32;
+
+/// A platform interface
+pub trait System {
+    /// Allocates a memory region of `size` bytes
+    unsafe fn alloc(size: usize) -> (*mut u8, usize, u32);
+
+    /// Remaps a memory region
+    unsafe fn remap(ptr: *mut u8, oldsize: usize, newsize: usize, can_move: bool) -> *mut u8;
+
+    /// Frees a part of a memory region
+    unsafe fn free_part(ptr: *mut u8, oldsize: usize, newsize: usize) -> bool;
+
+    /// Frees an entire memory region
+    unsafe fn free(ptr: *mut u8, size: usize) -> bool;
+
+    /// Indicates if the platform can release a part of memory
+    fn can_release_part(flags: u32) -> bool;
+
+    /// Allocates a memory region of zeros
+    fn allocates_zeros() -> bool;
+
+    /// Returns the page size
+    fn page_size() -> usize;
+}
+
+/// An allocator instance
+///
+/// Instances of this type are used to allocate blocks of memory. For best
+/// results only use one of these. Currently doesn't implement `Drop` to release
+/// lingering memory back to the OS. That may happen eventually though!
+#[cfg(any(target_os = "linux", target_arch = "wasm32", target_os = "macos"))]
+pub struct Dlmalloc<S = Platform>(dlmalloc::Dlmalloc<S>);
+
 /// An allocator instance
 ///
 /// Instances of this type are used to allocate blocks of memory. For best
 /// results only use one of these. Currently doesn't implement `Drop` to release
 /// lingering memory back to the OS. That may happen eventually though!
-pub struct Dlmalloc(dlmalloc::Dlmalloc);
+#[cfg(not(any(target_os = "linux", target_arch = "wasm32", target_os = "macos")))]
+pub struct Dlmalloc<S>(dlmalloc::Dlmalloc<S>);
 
 /// Constant initializer for `Dlmalloc` structure.
-pub const DLMALLOC_INIT: Dlmalloc = Dlmalloc(dlmalloc::DLMALLOC_INIT);
+#[cfg(any(target_os = "linux", target_os = "macos", target_arch = "wasm32"))]
+pub const DLMALLOC_INIT: Dlmalloc<Platform> = Dlmalloc::new();
 
 #[cfg(target_arch = "wasm32")]
 #[path = "wasm.rs"]
 mod sys;
 
+#[cfg(target_arch = "wasm32")]
+pub use sys::Wasm32;
+
 #[cfg(target_os = "macos")]
 #[path = "macos.rs"]
 mod sys;
 
+#[cfg(target_os = "macos")]
+pub use sys::MacOS;
+
 #[cfg(target_os = "linux")]
 #[path = "linux.rs"]
 mod sys;
 
-#[cfg(target_env = "sgx")]
-#[path = "sgx.rs"]
-mod sys;
+#[cfg(target_os = "linux")]
+pub use sys::Linux;
```

Comment:
Instead of this and
```rust
/// Type alias for the current platform
#[cfg(target_os = "linux")]
pub type Platform = Linux;
```

Can you just name the relevant type Platform in the linux module? That avoid a lot of cfged code